### ISSUE TITLE
Options/optimizations to activate when building opencv

### ISSUE THREAD

**Issue by sdvillal** [loopbio/opencv-feedstock#10]
Please @nzjrs complete here anytime with whatever you feel like:

- ```-DWITH_OPENCL=ON```: [nah](https://opencv.org/platforms/opencl.html); anything is exposed to python?

- ```-DWITH_OPENGL=ON```: nah, or disable to fix https://github.com/loopbio/conda-environments/issues/10

- ```-DWITH_TBB=ON```: we should do; for conda TBB is in conda-forge, defaults and intel channels

- ```-DWITH_IPP=ON```: full-blown IPP is up in intel channel (so I guess now it is free of charge?), or we could keep using the "free" version; ~~not so many opencv operations use it, and so it is better to call directly the counterpart in IPP if it exists (but we won't be doing that anytime soon, I guess)~~. See discussion on IPP down.

- ```-DWITH_CUDA=ON```: we already did, very detailed (see [old branch](https://github.com/loopbio/opencv-feedstock/tree/pre201708-loopbio-turbo-cuda)); not really accessible from python, it increases enormously compilation times and package size (even if we restrict the cuda capabilities we compile); if this is ever interesting, it can be done easily (make sure we compile against the conda cuda package); nvcuvid/nvdec could potentially be useful to read video directly into the GPU and maybe feed faster our models (although that is a long shot).

- ```-DENABLE_FAST_MATH=ON``` I would do this [in spite of the risk](https://stackoverflow.com/questions/7420665/what-does-gccs-ffast-math-actually-do)

- ```-march=native```: could be useful to avoid overall the cost of dynamic dispatching / making sure everything that can be optimized is optimized

**Comment by nzjrs** [comment]
> -DWITH_OPENCL=ON: nah

is nah what you have set or what you suggest? I don't care about OPENCL

> -D WITH_OPENGL=ON: nah, or disable to fix loopbio/conda-environments#10

don't know what uses this. done care. I guess `=OFF` is what we want

> -D WITH_TBB=ON: we should do, TBB is in conda-forge

I want this ON

> -D WITH_IPP=ON: full-blown IPP is up in intel channel (so I guess now it is free of charge?), 

Normally when you build opencv it downloads the compatible version and builds against that. That particular version made available in this way has a licence that makes it now un-problematic. Don't touch the separate one in the intel channel if you aren't sure about the license on that distribution. Use the one that OpenCV downloads so it may use it.

> or we 
> could keep using the "free" version; 

Free, not "free" as per my understanding and how it was the last time I checked

> not so many opencv operations use it, 

this isn't true. It's used extensively in opencv still AIUI (last time I checked = 3.1 and it was used in many core implementations)

> and so it is better to call directly the counterpart in IPP if it exists 

Nope. OpenCV delegates to IPP if enabled for many hot paths. 

> (but we won't be doing that anytime 
> soon, I guess)

no need. OpenCV delegates for us.

> -DWITH_CUDA=ON: we already did, very detailed (see old branch); not really accessible from python, 
> it increases enormously compilation times and package size (even if we restrict the cuda capabilities 
> we compile); if this is ever interesting, it can be done easily (make sure we compile against the conda 
> cuda package)

probbably safer to have this off considerting when I install a loopy instance with a GPU then I must install the nvidia libcuda

> -march=native: could be useful to avoid overall the cost of dynamic dispatching / making sure 
> everything that can be optimized is optimized

I think this is a red-herring

**Comment by sdvillal** [comment]
Thanks for commenting. I will incrementally apply all your suggestions.

nah means I do not think I should care...

I have no clue about opencv optimisation (the IPP blah is from reading old random posts in forums, that is actually how I have learned everything I know... the only tests I ever made is jpeg-turbo vs non-jpeg-turbo, it matters a lot when one keeps compressing frames from videos, and that is why I will eventually become bald).

**Comment by sdvillal** [comment]
Out of curiosity, this is the license for IPP

[LICENSE.txt](https://github.com/loopbio/opencv-feedstock/files/1687248/LICENSE.txt)

2-clause BSD heavily modified with "no reverse engineering", "no patents based on this software"... and a free ticket to Delaware!

**Comment by nzjrs** [comment]
This is the IPP licence for the mini version that OpenCV build process downloads? or the full license? Prior to opencv 3.1 the licences were radically different.

> I have no clue about opencv optimisation (the IPP blah is from reading old random posts in forums, 
> that is actually how I have learned everything I know

you should ignore everything you read. IPP was cool in the 2.X days, then got less cool between 2.X and 3.X due to licences, then got cool again in 3.X as it is now distributable. It is used extensively in OpenCV but it is hard to follow and stackoverflow posts are in general wrong because since 2.X and even once in the 3.X series OpenCV has rewritten they hardware/performance abstraction layer several times.

**Comment by sdvillal** [comment]
That is the license of the [newish IPP package in conda](https://anaconda.org/intel/ipp). Although it also talks about TBB in there, we can get TBB with either that commercial-maybe license or apache 2 license. I would not bother right now, but we should keep in mind that, a priori, it is very easy to test if we get gains using these libraries. It should be easy also to keep a testing ```opencv=3.4=nonfree*``` package where we put these things - once I'm happy again with the opencv package.

I have a quick question. Do you have an idea if BLAS is used anywhere in opencv, apart from the DNN module? The release notes for 3.3+ says that compiling against BLAS is not advantageous anymore in there.

**Comment by nzjrs** [comment]
I don't know if that is the licence that is applied to the version of IPP that is downloaded by OpenCV in the build process. I also don't want to use IPP in conda just because it is in conda. The magic OpenCV only one is good enough for me. Everything we seem to use something 'because it is in conda' it seems to baloon into a giant fractal ball of failure ;-)

I have tested gains using IPP in 3.1/3.2 it was huge for many hot paths (resize, colorspace converstion, grayscale converstion, data type changes, etc).

No idea about BLAS I'm afraid.

Remember, there are *extensive* performance tests in python-loopb if you are curious about these things

**Comment by sdvillal** [comment]
I imagine the (how much castrated?) IPP version downloaded by opencv should have a more standard license.

**Comment by sdvillal** [comment]
Intel info: https://software.intel.com/en-us/articles/free-mkl

> Community Licensing for Intel® Performance Libraries – free for all, registration required, no royalties, no restrictions on company or project size, current versions of libraries, no Intel Premier Support access.

**Comment by nzjrs** [comment]
What is the current relationship of MKL to BLAS to IPP to OpenCV? If you use MKL as your BLAS provider then does that impact IPP use? It used to be completely separate - i.e. my build was OpenBLAS and IPP with no mention of MLK (because at that time it was not freely usable).

Sincle Intel cuddles with Anaconda I think they have changed some licenses.

**Comment by sdvillal** [comment]
They are separated, of course, one abides only by the license of what is installed. I guess they are all now under the same umbrella, as the article I sent talks about all MKL, IPP, TBB. Not only for conda, they have just gone full free. It only happens they use their conda channel to make it easier for people to grab it.

Is there any difference between what is in the channel and the version that is [automatically downloaded by opencv](https://github.com/opencv/opencv/blob/9c14a2f0aa148851923b4edc1d7f4fd6ffe16d20/3rdparty/ippicv/ippicv.cmake)? ([this one as of today](https://raw.githubusercontent.com/opencv/opencv_3rdparty/dfe3162c237af211e98b8960018b564bc209261d/ippicv/ippicv_2017u3_lnx_intel64_general_20170822.tgz)). Not in terms of license, the one in opencv seems the same to me as the one I linked a couple of posts ago. I cannot tell about functionality, as what comes in the conda version strikes to me as a superset of what opencv downloads (a single .a and stripped down headers) if not provided with a location to IPP. Would opencv use anything different if we provide it? No idea, but I doubt so.

MKL is now in anaconda defaults, mirroring somehow the package in the intel channel (but compiled with the gcc suite). I think we are very fine using any of these. And I would very much ditch compiling opencv against any BLAS if, as I suspect, it would make no difference in performance to us. So time for some loopb benchmarks...

**Comment by sdvillal** [comment]
So TL;DR

- We should be OK using IPP, TBB and MKL.
- For opencv, I would keep IPP to the one provided by opencv.
- I would just not compile it against any blas, just to save any trouble. But needs benchmarking (or someone who knows telling us I'm right for this one).

**Comment by nzjrs** [comment]
> I cannot tell about functionality, as what comes in the conda version strikes to me as a superset of 
> what opencv downloads (a single .a and stripped down headers) if not provided with a location to IPP. 
> Would opencv use anything different if we provide it? No idea, but I doubt so.

Not to mention by distributing the `.a` and headers it makes compilation way easier as they dont have to ship or deal with IPP version changes / system library versions.

> Would opencv use anything different if we provide it? No idea, but I doubt so.

It seems to be that opencv will continue to use the bundled one regardless of the presence of IPP on the system if you say all they download is the `.a`

> BLAS if, as I suspect, it would make no difference in performance to us. So time for some loopb 
> benchmarks

If you can think of something that definately uses BLAS then I can also write a performance test.

**Comment by sdvillal** [comment]
That is until they put all the symbols of the .a in a shared library and they screw the world for everyone else using a non-compatible version of IPP (lucklily nothing else!) ;-).

The only think I know for sure that can use BLAS is the DNN module. But they stress in the 3.3+ release notes that their code there now goes as fast as with BLAS. Besides, we do not plan to use DNN. Do you remember compiling against BLAS before the DNN module was a thing?

**Comment by nzjrs** [comment]
> Besides, we do not plan to use DNN. Do you remember compiling against BLAS before the DNN 
> module was a thing?

grepping the source code tells me the HAL layer still has entries for LAPACK which should delegate to BLAS (or implementations?). Of note is there is an entry point `cv::gemm()` and `cv2.gemm()`. So I guess you could write a gemm test and see if/where LAPACK delegates to, and if it's BLAS or if the HAL picks it up and has an internal (depending on the build) `gemm()`.

Outside of gemm maybe using BLAS eventually I can't another 'direct' BLAS  call in the code.

**Comment by nzjrs** [comment]
> That is until they put all the symbols of the .a in a shared library and they screw the world for everyone 
> else using a non-compatible version of IPP (lucklily nothing else!) ;-).

I think that's a misdiagnosis of the libjpeg crap or a bad build step elsewhere. AFAIUI GCC/LD can/and by default do? permute symbols in an `.a` when linking it statically into another build. Wasn't the libjpeg turbo thing two libraries claiming to be the same ABI but not actually being so? explicitly passing the libjpeg.so to opencv should fix this by my understanding.

**Comment by sdvillal** [comment]
I believe there is no misdiagnosis in the libjpeg case - but of course it does not need to apply here, it was just a humorous license). 

I will write a quick summary in the other issue so you can understand what I understand - and correct me wrong if you need. I do not think you need to waste any brain cycle on it. But quick points here: I know a workaround to have all the cake (speed and no segfaults) but to me that is not a proper fix. libjpeg is C, no symbol mangling. Once you load the dynamic symbols from a version of the library, they stay there for whoever needs them. Opencv exports libjpeg names in its codecs module when statically compiling libjpeg-turbo... will follow the argument.

**Comment by nzjrs** [comment]
I'm with you on the diagnosis so far.

**Comment by sdvillal** [comment]
Some links from the opencv wiki:
- [OpenCV 3 & IPP](https://github.com/opencv/opencv/wiki/Opencv3#intel%EF%B8%8F-ipp)
- [CPU optimizations](https://github.com/opencv/opencv/wiki/CPU-optimizations-build-options)
- [OpenCL optimizations](https://github.com/opencv/opencv/wiki/CPU-optimizations-build-options)
- [Compact build advice](https://github.com/opencv/opencv/wiki/Compact-build-advice)

**Comment by sdvillal** [comment]
06e8f32  selected optimizations ON, others explicitly OFF (ready to experiment in a "experimental" version of the package).
